pilot/pkg/xds/eds.go

```diff
@@ -183,36 +187,43 @@ func (s *DiscoveryServer) edsCacheUpdate(shard model.ShardKey, hostname string,
 		Namespace: namespace,
 	}: {}})
 	ep.mutex.Unlock()
-	if features.SendUnhealthyEndpoints && pushType == IncrementalPush {
+	if features.SendUnhealthyEndpoints {
 		// Check if new Endpoints are ready to be pushed. This check
 		// will ensure that if a new pod comes with a non ready endpoint,
 		// we do not unnecessarily push that config to Envoy.
 		// Please note that address is not a unique key. So this may not accurately
 		// identify based on health status and push too many times - which is ok since its an optimization.
 		emap := make(map[string]*model.IstioEndpoint, len(oldIstioEndpoints))
+		// Add new endpoints only if they are ever ready once to shards
+		// so that full push does not send them from shards.
+		newIstioEndpoints := []*model.IstioEndpoint{}
 		for _, oie := range oldIstioEndpoints {
 			emap[oie.Address] = oie
 		}
 		needPush := false
 		for _, nie := range istioEndpoints {
 			if oie, exists := emap[nie.Address]; exists {
 				// If endpoint exists already, we should push if it's health status changes.
-				needPush = oie.HealthStatus != nie.HealthStatus
-			} else {
+				if oie.HealthStatus != nie.HealthStatus {
+					needPush = true
+				}
+				newIstioEndpoints = append(newIstioEndpoints, nie)
+			} else if nie.HealthStatus == model.Healthy {
 				// If the endpoint does not exist in shards that means it is a
 				// new endpoint. Only send if it is healthy to avoid pushing endpoints
 				// that are not ready to start with.
-				needPush = nie.HealthStatus == model.Healthy
-			}
-			if needPush {
-				break
+				needPush = true
+				newIstioEndpoints = append(newIstioEndpoints, nie)
 			}
 		}
 
-		if !needPush {
+		if pushType != FullPush && !needPush {
 			log.Infof("No push, either old endpoint health status did not change or new endpoint came with unhealthy status, %v", hostname)
 			pushType = NoPush
 		}
+		ep.mutex.Lock()
+		ep.Shards[shard] = newIstioEndpoints
+		ep.mutex.Unlock()
```

Comment:
I am not sure why its safe to set `ep.Shards[shard]` twice - here and L167, with different values? Can't the 'wrong' value be used in the mean time?

I would expect we set it once

Reply:
Yeah. I did not think about that case. Thanks for catching it. Fixed it now.